File: DemoServer/AdditionalLanguages/go/RelatedDocuments/indexRelatedDocuments/IndexRelatedDocuments.go
package main

//region Usings
import "github.com/ravendb/ravendb-go-client"
//endregion

var globalDocumentStore *ravendb.DocumentStore

func main() {
    createDocumentStore()
    createDatabase()
    indexRelatedDocuments("Produce")
    globalDocumentStore.Close()
}

func createDocumentStore() (*ravendb.DocumentStore, error) {
    if globalDocumentStore != nil {
        return globalDocumentStore, nil
    }
    urls := []string{"http://localhost:8080"}
    store := ravendb.NewDocumentStore(urls, "testGO")
    err := store.Initialize()
    if err != nil {
        return nil, err
    }
    globalDocumentStore = store
    return globalDocumentStore, nil
}

func createDatabase() {
    databaseRecord := ravendb.NewDatabaseRecord()
    databaseRecord.DatabaseName = "testGO"
    createDatabaseOperation := ravendb.NewCreateDatabaseOperation(databaseRecord, 1)
    var err = globalDocumentStore.Maintenance().Server().Send(createDatabaseOperation)
    if err != nil {
        fmt.Printf("d.store.Maintenance().Server().Send(createDatabaseOperation) failed with %s\n", err)
    }
}

//region Demo
func indexRelatedDocuments(categoryName string) error {
    
    //region Step_1
    index := ravendb.NewIndexCreationTask("Products/ByCategoryName")
    //endregion
    
    //region Step_2
    index.Map = `docs.Products.Select(product => new {
        CategoryName = (this.LoadDocument(product.Category, "Categories")).Name
    })`
    //endregion
    
    //region Step_3
    err := globalDocumentStore.ExecuteIndex(index, "")
    if err != nil {
        return err
    }
    //endregion
    
    session, err := globalDocumentStore.OpenSession("")
    if err != nil {
        return err
    }
    defer session.Close()

    //region Step_4  
    q := session.QueryIndex(index.IndexName)
    q = q.WhereEquals("CategoryName", categoryName)
    
    var productsWithCategoryName []*Product
    err = q.GetResults(&productsWithCategoryName)
    //endregion
    
    if err != nil {
        return err
    }
    return nil
}

type Product struct {
    ID   string
    Name string
    Supplier string
    Category string
}
//endregion

File: DemoServer/AdditionalLanguages/java/src/main/java/net/ravendb/demo/relatedDocuments/indexRelatedDocuments/IndexRelatedDocuments.java
package net.ravendb.demo.relatedDocuments.indexRelatedDocuments;

//region Usings
import net.ravendb.client.documents.indexes.AbstractIndexCreationTask;
import net.ravendb.client.documents.session.IDocumentSession;
import java.util.List;
//endregion
import net.ravendb.demo.common.DocumentStoreHolder;
import net.ravendb.demo.common.models.Product;

public class IndexRelatedDocuments {

    //region Demo
    //region Step_1
    public static class Products_ByCategoryName extends AbstractIndexCreationTask {
    //endregion

        //region Step_2
        public Products_ByCategoryName() {
            map = "docs.products.Select(product => new { " +
                "    categoryName = (this.LoadDocument(product.Category, \"Categories\")).Name " +
                "})";
        }
        //endregion
    }
    //endregion

    public List<Product> run(RunParams runParams) {
        String categoryName = runParams.getCategoryName();

        //region Demo
        List<Product> productsWithCategoryName;

        try (IDocumentSession session = DocumentStoreHolder.store.openSession()) {
            //region Step_3
            productsWithCategoryName = session
                .query(Product.class, Products_ByCategoryName.class)
                .whereEquals("categoryName", categoryName)
                .toList();
            //endregion
        }
        //endregion

        return productsWithCategoryName;
    }

    public static class RunParams {
        private String categoryName;

        public String getCategoryName() {
            return categoryName;
        }

        public void setCategoryName(String categoryName) {
            this.categoryName = categoryName;
        }
    }
}
